### PR TITLE
Implement "feefilter" P2P message

### PR DESCRIPTION
**Pull request by morcos**
Please see mailing list discussion
https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2016-February/012449.html

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -5146,6 +5162,19 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
         }
     }
 
+    else if (strCommand == NetMsgType::FEEFILTER)
+    {
+        CAmount newFeeFilter = 0;
+        vRecv >> newFeeFilter;
+        if (MoneyRange(newFeeFilter)) {
+            {
+            LOCK(pfrom->cs_feeFilter);
```
Comment (jonasschnelli) [inline src/main.cpp]: Nit: indent.

**Comment by jonasschnelli** [comment]
Nice Work!
IMO relatively important PR to reduce p2p "noise".
Concept ACK. Short code review.

File: src/main.cpp [hunk 2]
```diff
@@ -5146,6 +5162,19 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
         }
     }
 
+    else if (strCommand == NetMsgType::FEEFILTER)
+    {
+        CAmount newFeeFilter = 0;
+        vRecv >> newFeeFilter;
+        if (MoneyRange(newFeeFilter)) {
+            {
```
Comment (arowser) [inline src/main.cpp]: Why add two  "{" in here?

Comment (jonasschnelli) [inline src/main.cpp]: To unlock the thread-lock for the LogPrint at L5174.

File: src/main.cpp [hunk 3]
```diff
@@ -5629,6 +5658,29 @@ bool SendMessages(CNode* pto)
         if (!vGetData.empty())
             pto->PushMessage(NetMsgType::GETDATA, vGetData);
 
+        //
+        // Message: feefilter
+        //
+        // We want don't want white listed peers to filter txs to us if we have -whitelistforcerelay
```
Comment (paveljanik) [inline src/main.cpp]: We want don't want?

File: src/main.h [hunk 4]
```diff
@@ -120,6 +124,8 @@ static const bool DEFAULT_ENABLE_REPLACEMENT = true;
 
 /** Maximum number of headers to announce when relaying blocks with headers message.*/
 static const unsigned int MAX_BLOCKS_TO_ANNOUNCE = 8;
+/** Default for using fee filter*/
```
Comment (paveljanik) [inline src/main.h]: SPC at the end missing

File: src/policy/fees.h [hunk 5]
```diff
@@ -286,4 +286,17 @@ class CBlockPolicyEstimator
     CFeeRate feeLikely, feeUnlikely;
     double priLikely, priUnlikely;
 };
+
+class FeeFilterRounder
+{
+public:
+    /** Create new FeeFilterRounder*/
```
Comment (paveljanik) [inline src/policy/fees.h]: SPC at the end of comment

**Comment by paveljanik** [comment]
Any volunteer to write some test cases?

File: src/protocol.h [hunk 6]
```diff
@@ -218,7 +218,12 @@ extern const char *REJECT;
  * @see https://bitcoin.org/en/developer-reference#sendheaders
  */
 extern const char *SENDHEADERS;
-
+/**
+ * The feefilter message tells the receiving peer not to inv us any txs
+ * which do not meet the specified min fee rate.
+ * @since protocol version 70013 as described by BIPXXX
```
Comment (paveljanik) [inline src/protocol.h]: @luke-jr Can you please assign a BIP number?

**Comment by paveljanik** [comment]
What will we do when we sent `feefilter` to the peer and the peer sends us zero fee `inv`s?

File: src/main.cpp (hunk 1 above)
Comment (MarcoFalke) [inline src/main.cpp]: I remember adding the clang-format script some time ago. Something like `git diff -U0 HEAD~1.. | ./contrib/devtools/clang-format-diff.py -p1 -i -v` should do all of this.

File: src/protocol.h (hunk 6 above)
Comment (luke-jr) [inline src/protocol.h]: @morcos needs to request this.

**Comment by morcos** [comment]
Addressed the cleanup comments
This has been submited as a [PR](https://github.com/bitcoin/bips/pull/341) for BIP 133.

@paveljanik It's impossible to be sure there aren't invs or txs in flight with fee rates below the cut off. Also in general we don't have a good system for responding to p2p misbehavior.  I think this could be extended later with that ability, but as of now it does not appear to open up any additional DoS attacks if you don't abide by the message.

File: src/main.cpp [hunk 7]
```diff
@@ -5629,6 +5653,29 @@ bool SendMessages(CNode* pto)
...
+            int64_t timeNow = GetTimeMicros();
+            if (timeNow > pto->nextSendTimeFeeFilter) {
+                CAmount filterToSend = filterRounder.round(currentFilter);
+                if (filterToSend != pto->lastSentFeeFilter) {
+                    pto->PushMessage(NetMsgType::FEEFILTER, filterToSend);
+                    pto->lastSentFeeFilter = filterToSend;
+                }
+                pto->nextSendTimeFeeFilter = PoissonNextSend(timeNow, AVG_FEEFILTER_BROADCAST_INTERVAL);
+            }
+            // If the fee filter has changed substantially and its still more than MAX_FEEFILTER_CHANGE_DELAY
```
Comment (paveljanik) [inline src/main.cpp]: its -> it is/it's

File: src/wallet/wallet.cpp [hunk 8]
```diff
@@ -1265,9 +1265,11 @@ bool CWalletTx::RelayWalletTransaction()
     assert(pwallet->GetBroadcastTransactions());
     if (!IsCoinBase())
     {
-        if (GetDepthInMainChain() == 0 && !isAbandoned()) {
+        if (GetDepthInMainChain() == 0 && !isAbandoned() && InMempool()) {
```
Comment (paveljanik) [inline src/wallet/wallet.cpp]: This have been merged now - can you please rebase?

**Comment by morcos** [comment]
Squashed and rebased

**Comment by MarcoFalke** [comment]
Please mention the bip number in the github subject line. Also you'd need to update `/doc/bips.md` prior to merge.

File: doc/release-notes.md [hunk 9]
```diff
@@ -51,6 +51,14 @@ The following outputs are affected by this change:
 
 ### P2P protocol and network code
 
+Fee filtering of invs (BIP 133)
```
Comment (MarcoFalke) [inline doc/release-notes.md]: Nit: I am pretty sure this goes up to "Notable changes" (`Example item`) as this place down here is usually used for the detailed change log only.

File: src/main.cpp [hunk 10]
```diff
@@ -938,7 +941,7 @@ std::string FormatStateMessage(const CValidationState &state)
 }
 
 bool AcceptToMemoryPoolWorker(CTxMemPool& pool, CValidationState& state, const CTransaction& tx, bool fLimitFree,
-                              bool* pfMissingInputs, bool fOverrideMempoolLimit, const CAmount nAbsurdFee,
+                              bool* pfMissingInputs, CAmount* txFeeRate, bool fOverrideMempoolLimit, const CAmount nAbsurdFee,
```
Comment (MarcoFalke) [inline src/main.cpp]: Nit: can you add the missing ampersand to `CAmount nAbsurdFee` as well here?

File: src/net.cpp [hunk 11]
```diff
@@ -2063,20 +2063,15 @@ class CNetCleanup
 instance_of_cnetcleanup;
 
 
-
-
-
-
-
-void RelayTransaction(const CTransaction& tx)
+void RelayTransaction(const CTransaction& tx, CAmount fee)
```
Comment (MarcoFalke) [inline src/net.cpp]: `CAmount fee` implies an absolute fee. Maybe `CAmount feeRate`?

Personally, I'd prefer to use `CFeeRate` in places except `CNode`. I.e. something like this should suffice:

```diff
diff --git a/src/main.cpp b/src/main.cpp
index 487564a..c0fccbf 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -944 +944 @@ bool AcceptToMemoryPoolWorker(CTxMemPool& pool, CValidationState& state, const C
-                              bool* pfMissingInputs, CAmount* txFeeRate, bool fOverrideMempoolLimit, const CAmount nAbsurdFee,
+                              bool* pfMissingInputs, CFeeRate* txFeeRate, bool fOverrideMempoolLimit, const CAmount nAbsurdFee,
@@ -1101 +1101 @@ bool AcceptToMemoryPoolWorker(CTxMemPool& pool, CValidationState& state, const C
-            *txFeeRate = CFeeRate(nFees, nSize).GetFeePerK();
+            *txFeeRate = CFeeRate(nFees, nSize);
@@ -1362 +1362 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
-                        bool* pfMissingInputs, CAmount* txFeeRate, bool fOverrideMempoolLimit, const CAmount nAbsurdFee)
+                        bool* pfMissingInputs, CFeeRate* txFeeRate, bool fOverrideMempoolLimit, const CAmount nAbsurdFee)
@@ -4829 +4829 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
-        CAmount txFeeRate = 0;
+        CFeeRate txFeeRate = CFeeRate(0);
@@ -4863 +4863 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
-                    CAmount orphanFeeRate = 0;
+                    CFeeRate orphanFeeRate = CFeeRate(0);
diff --git a/src/main.h b/src/main.h
index 5c006b2..fe4abb5 100644
--- a/src/main.h
+++ b/src/main.h
@@ -293 +293 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
-                        bool* pfMissingInputs, CAmount* txFee, bool fOverrideMempoolLimit=false, const CAmount nAbsurdFee=0);
+                        bool* pfMissingInputs, CFeeRate* txFee, bool fOverrideMempoolLimit=false, const CAmount nAbsurdFee=0);
diff --git a/src/net.cpp b/src/net.cpp
index 294d2d4..22e8b26 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -2066 +2066 @@ instance_of_cnetcleanup;
-void RelayTransaction(const CTransaction& tx, CAmount fee)
+void RelayTransaction(const CTransaction& tx, CFeeRate fee)
@@ -2074 +2074 @@ void RelayTransaction(const CTransaction& tx, CAmount fee)
-void RelayTransaction(const CTransaction& tx, CAmount fee, const CDataStream& ss)
+void RelayTransaction(const CTransaction& tx, CFeeRate fee, const CDataStream& ss)
@@ -2097 +2097 @@ void RelayTransaction(const CTransaction& tx, CAmount fee, const CDataStream& ss
-            if (fee < pnode->minFeeFilter)
+            if (fee.GetFeePerK() < pnode->minFeeFilter)
diff --git a/src/net.h b/src/net.h
index 0b3400b..3d4117d 100644
--- a/src/net.h
+++ b/src/net.h
@@ -776,2 +776,2 @@ class CTransaction;
-void RelayTransaction(const CTransaction& tx, CAmount fee);
-void RelayTransaction(const CTransaction& tx, CAmount fee, const CDataStream& ss);
+void RelayTransaction(const CTransaction& tx, CFeeRate fee);
+void RelayTransaction(const CTransaction& tx, CFeeRate fee, const CDataStream& ss);
diff --git a/src/rpc/rawtransaction.cpp b/src/rpc/rawtransaction.cpp
index 5c93846..931eaeb 100644
--- a/src/rpc/rawtransaction.cpp
+++ b/src/rpc/rawtransaction.cpp
@@ -821 +821 @@ UniValue sendrawtransaction(const UniValue& params, bool fHelp)
-    CAmount txFee = 0;
+    CFeeRate txFee = CFeeRate(0);
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 586fc69..0070ad1 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -1270 +1270 @@ bool CWalletTx::RelayWalletTransaction()
-            RelayTransaction((CTransaction)*this, feeRate.GetFeePerK());
+            RelayTransaction((CTransaction)*this, feeRate);

**Comment by MarcoFalke** [comment]
Concept ACK  ef3b850

File: src/net.cpp (hunk 11 above)
Comment (morcos) [inline src/net.cpp]: @MarcoFalke I wasn't sure about that.
I didn't love the idea of introducing CFeeRate into net.cpp and RelayTransaction, but if people prefer it that way I don't have a strong opinion.

**Comment by morcos** [comment]
@MarcoFalke OK I made your suggested change

File: src/net.cpp (hunk 11 above)
Comment (MarcoFalke) [inline src/net.cpp]: I mostly care about having it in the mempool, so the code can be reused by other logic than just feefilter.

Thanks for addressing the nits.

File: qa/rpc-tests/maxuploadtarget.py [hunk 12]
```diff
@@ -7,7 +7,6 @@
 from test_framework.mininode import *
 from test_framework.test_framework import BitcoinTestFramework
 from test_framework.util import *
-from test_framework.comptool import wait_until
```
Comment (paveljanik) [inline qa/rpc-tests/maxuploadtarget.py]: Why is this removed in this PR?

Comment (paveljanik) [inline qa/rpc-tests/maxuploadtarget.py]: Ah you moved it to mininode. OK.

File: src/main.h [hunk 13]
```diff
@@ -284,7 +290,7 @@ void PruneAndFlush();
 
 /** (try to) add transaction to memory pool **/
 bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransaction &tx, bool fLimitFree,
-                        bool* pfMissingInputs, bool fOverrideMempoolLimit=false, const CAmount nAbsurdFee=0);
+                        bool* pfMissingInputs, CFeeRate* txFeeRate, bool fOverrideMempoolLimit=false, const CAmount nAbsurdFee=0);
```
Comment (MarcoFalke) [inline src/main.h]: tiny nit: Add missing ampersand to `CAmount nAbsurdFee`?

**Comment by sdaftuar** [comment]
ACK

**Comment by paveljanik** [comment]
ACK

File: src/main.cpp [hunk 14]
```diff
@@ -2498,7 +2504,7 @@ bool static DisconnectTip(CValidationState& state, const Consensus::Params& cons
         // ignore validation errors in resurrected transactions
         list<CTransaction> removed;
         CValidationState stateDummy;
-        if (tx.IsCoinBase() || !AcceptToMemoryPool(mempool, stateDummy, tx, false, NULL, true)) {
+        if (tx.IsCoinBase() || !AcceptToMemoryPool(mempool, stateDummy, tx, false, NULL, NULL, true)) {
             mempool.remove(tx, removed, true);
```
Comment (MarcoFalke) [inline src/main.cpp]: Needs merge conflict resolved.

**Comment by morcos** [comment]
Squashed 1 commit and rebased to address merge conflict (trivial)

**Comment by laanwj** [comment]
utACK 0371797

File: src/main.cpp [hunk 15]
```diff
@@ -5845,6 +5869,29 @@ bool SendMessages(CNode* pto)
...
             pto->PushMessage(NetMsgType::GETDATA, vGetData);
 
+        //
+        // Message: feefilter
+        //
+        // We don't want white listed peers to filter txs to us if we have -whitelistforcerelay
+        if (pto->nVersion >= FEEFILTER_VERSION && GetBoolArg("-feefilter", DEFAULT_FEEFILTER) &&
+            !(pto->fWhitelisted && GetBoolArg("-whitelistforcerelay", DEFAULT_WHITELISTFORCERELAY))) {
+            CAmount currentFilter = mempool.GetMinFee(GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000).GetFeePerK();
+            int64_t timeNow = GetTimeMicros();
```
Comment (rebroad) [inline src/main.cpp]: why not use nNow variable which is also set to GetTimeMicros() earlier in this function?
